skymap_scanner/utils/extract_json_message.py

```diff
@@ -135,85 +134,77 @@ def __extract_frame_packet(
+        baseline_GCD_file = os.path.join(GCD_dir, baseline_GCD)
+        LOGGER.debug(f"Trying GCD file: {baseline_GCD_file}")
+        if not os.path.isfile(baseline_GCD_file):
+            raise RuntimeError("Baseline GCD file not available!")
+        # NOTE: logic allowing GCD_dir to point to a file, in order to directly override the GCD has been removed.
+        
+        cached_baseline_GCD_file = os.path.join(event_cache_dir, cfg.BASELINE_GCD_FILENAME)
 
-    # special case for old EHE alerts with empty GCD frames
-    if ("I3Geometry" not in frame_packet[0]) and ("I3GeometryDiff" not in frame_packet[0]):
+        baseline_GCD_framepacket = load_framepacket_from_file(baseline_GCD_file)
 
-        # If no GCD is specified, work out correct one from run number
+        if os.path.exists(cached_baseline_GCD_file):
+            # this should occur if the cache is isolated on a server-instance basis
+            # but we keep it for the time being in case we want to read back an old scan
+            baseline_GCD_hash = hash_frame_packet(baseline_GCD_framepacket)
 
-        available_GCDs = sorted([x for x in os.listdir(override_GCD_filename) if ".i3" in x])
+            cached_baseline_GCD_framepacket = load_framepacket_from_file(cached_baseline_GCD_file)
+            cached_baseline_GCD_hash = hash_frame_packet(cached_baseline_GCD_framepacket)
+
+            if cached_baseline_GCD_hash != baseline_GCD_hash:
+                raise RuntimeError(f"Existing baseline GCD in cache (SHA1 {cached_baseline_GCD_hash}) and packet from input (SHA1 {baseline_GCD_hash}) differ.")
+            LOGGER.debug("Checked baseline GCD against existing data in cache: consistent.")
+        else:
+            save_GCD_frame_packet_to_file(baseline_GCD_framepacket, cached_baseline_GCD_file)
+            LOGGER.debug(f"Wrote baseline GCD frames to {cached_baseline_GCD_file}.")
+
+        # baseline_GCD path is saved in a text file
+        source_baseline_GCD_metadata = os.path.join(event_cache_dir, cfg.SOURCE_BASELINE_GCD_METADATA)
+        if os.path.isfile(source_baseline_GCD_metadata):
+            with open(source_baseline_GCD_metadata, 'r') as f:
+                filename = f.read()
+            if (filename != baseline_GCD) and (os.path.basename(filename) != os.path.basename(baseline_GCD)):
+                raise RuntimeError(f"Expected the stored source baseline GCD to be {baseline_GCD}. It is {filename}.")
+        with open(source_baseline_GCD_metadata, "w") as text_file:
+            text_file.write(baseline_GCD)   
+
+    #=====================
+    # GCD-less framepacket
+    #=====================
+    if ("I3Geometry" not in frame_packet[0]) and ("I3GeometryDiff" not in frame_packet[0]):
+        LOGGER.debug("Packet with empty GCD frames. Need to load baseline GCD")
+        # If no GCD is specified, work out correct one from run number
+        available_GCDs = sorted([x for x in os.listdir(GCD_dir) if ".i3" in x])
```

Comment:
That's a lot of diff--I'll have to trust you here. Perhaps @tianluyuan could give a better review of this change.

Reply:
How does this actually pick out the GCD file for a particular GCD-less event? It looks like it will use the latest baseline GCD file that it can find, but that may not (probably does not) correspond to the actual GCD of the event?

I guess I am saying it seems like all frame packets should have at least a GCD-diff.

Reply:
Correct @tianluyuan . According to documentation this logic applies to old-format EHE alerts, but I found out it will also trigger if you pass as an input subthreshold GFU event (that is GCD-less). For the time being I think is good to support this scenario in the code (although with all possible warnings).